File: soundObjects/fracDelay.js
export function FracDelay(M, i) {
    this.maxDelay = 44100;
    this.interpolation = i;
    this.lastOut = 0;
    this.M = M > this.maxDelay ? this.maxDelay : M
    this.buffer = new Float32Array(this.maxDelay);
    this.pW = this.maxDelay;
    this.pR = this.pW - Math.floor(this.M);

    this.process = function (frame) {
        const wI = (this.pW % this.maxDelay) + 1;
        const rI = (this.pR % this.maxDelay) + 1;
        const rI2 = ((this.pR - 1) % this.maxDelay) + 1;
        const frac = this.M - Math.floor(this.M);
        let out = 0;
        this.buffer[wI] = frame;
        switch (i) {
            case 'Linear':
                out = frac * this.buffer[rI2] + (1 - frac) * this.buffer[rI];
                break;
            case 'AllPass':
                out = this.buffer[rI2] + ((1 - frac) * (this.buffer[rI] - this.lastOut));
                this.lastOut = isNaN(out) ? 0 : out;
                break;
        }

        this.pW = this.pW + 1;
        this.pR = this.pR + 1;
        const a = [this.pW, this.pR]

        // if(isNaN(out))debugger
        return isNaN(out) ? 0 : out;
    };

    this.setDelay = function (M) {
        this.M = M > this.maxDelay ? this.maxDelay : M;
        this.pR = this.pW - Math.floor(this.M);
    }
}

File: Implementaciones/fracDelay.m
classdef fracDelay<handle
    properties
        M
        maxDelay
        buffer
        pW
        pR
        interpolation
        lastOut
    end
    methods
        function obj=fracDelay(M,i)
            obj.maxDelay=44100;
            obj.interpolation=i;
            obj.lastOut=0;
            if M>obj.maxDelay   %Restriccion de delay maximo
                obj.M=obj.maxDelay;
            else
                obj.M=M;
            end
            obj.buffer=zeros(1,obj.maxDelay);
            obj.pW=obj.maxDelay;
            obj.pR=obj.pW-floor(obj.M);
        end
        function out=process(obj,x)
            wI=mod(obj.pW,obj.maxDelay)+1;
            rI=mod(obj.pR,obj.maxDelay)+1;
            rI2=mod(obj.pR-1,obj.maxDelay)+1;
            frac=obj.M-floor(obj.M);
            
            obj.buffer(wI)=x;
            switch obj.interpolation
                case 'Linear'
                    out=frac*obj.buffer(rI2) + (1-frac)*obj.buffer(rI); 
                case 'AllPass'
                    out=obj.buffer(rI2)+(1-frac)*(obj.buffer(rI)-obj.lastOut);
                    obj.lastOut=out;
            end
                                  
            obj.pW=obj.pW+1;
            obj.pR=obj.pR+1;
            a=[obj.pW;obj.pR]
        end
        function setDelay(obj,M)
            if M>obj.maxDelay   %Restriccion de delay maximo
                obj.M=obj.maxDelay;
            else
                obj.M=M;
            end
            obj.pR=obj.pW-floor(obj.M);
        end
    end
end